app/templates/chat-demo.html

```diff
@@ -1,56 +1,119 @@
             event.preventDefault();
-            if (ws.readyState == WebSocket.CLOSED || ws.readyState == WebSocket.CLOSING){
-              alert("Chat not connected. Did you give the correct token???");
-              location.reload()
+            if (ws.readyState == WebSocket.CLOSED || ws.readyState == WebSocket.CLOSING) {
+                alert("Chat not connected. Did you give the correct token???");
+                location.reload()
             }
             ws.send($('#message').val());
-            $('#messages').append('<div><b>You</b>: '+$('#message').val() + '</div>');
+            $('#messages').append('<div><b>You</b>: ' + $('#message').val() + '</div>');
             $('#message').val('').focus();
-          });
-        </script>
-      </body>
-      
+        });
+    </script>
+    <script>
+        let mediaRecorder;
+        let chunks = [];
+
+        // Check if the browser supports the required APIs
+        if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
+            navigator.mediaDevices.getUserMedia({ audio: true })
+                .then(function (stream) {
+                    const recordButton = document.getElementById("recordButton");
+                    let isRecording = false;
+
+                    recordButton.addEventListener("click", function () {
+                        if (mediaRecorder && mediaRecorder.state === "recording") {
+                            mediaRecorder.stop();
+                        } else {
+                            startRecording(stream);
+                        }
+                        // Toggle the button label
+                        isRecording = !isRecording;
+                        recordButton.textContent = isRecording ? "Stop" : "Record";
+                    });
+                })
+                .catch(function (error) {
+                    console.error("Error accessing microphone:", error);
+                });
+        } else {
+            console.error("getUserMedia not supported on your browser");
+        }
+
+        function startRecording(stream) {
+            mediaRecorder = new MediaRecorder(stream);
+            chunks = [];
+
+            mediaRecorder.addEventListener("dataavailable", function (event) {
+                chunks.push(event.data);
+            });
+
+            mediaRecorder.addEventListener("stop", function () {
+                const audioBlob = new Blob(chunks, { type: "audio/wav" });
+                const audioFile = new File([audioBlob], "recorded_audio.wav");
+                sendToSocket(audioFile);
+            });
+
+            mediaRecorder.start();
+        }
+
+        function sendToSocket(audioFile) {
+
+            const reader = new FileReader();
+            reader.onload = function (event) {
+                const arrayBuffer = event.target.result;
+                ws.send("audioData", arrayBuffer);
+            };
+            reader.readAsArrayBuffer(audioFile);
+        }
```

Comment:
The message passing to websocket is as follows now:
At UI: `ws.send($('#message').val());` , which is text
At server: ` question = await websocket.receive_text()`

So I can think of two ways to implement this connection
1. Implement whisper at backend: change the websocket code above to accept a JSON and have a field to indicate if its text or audio that is being passed and based on that process it at backend. Then the now empty core/auido/whisper.py will be called to convert the audio to text and passed to next step in websocket router function.
2. Implement Whisper process at the UI: Instead of sending audio to backend, make use of Whisper's JS APIs and perform the audio to text conversion at this JS script section itself and send text to back end. This way, audio support will not be a core feature of the server though, but we can still demo how it could be integrated as an addon.